index.js

```diff
@@ -0,0 +1,365 @@
+//.
+//. ```javascript
+//. > parseArgs (spec) (Pair (conf) (['--word', '--color', 'foo']))
+//. Right (Pair ({color: false, email: Nothing, words: ['--color']}) (['foo']))
+//. ```
+//.
+//. Option handlers may perform validation:
+//.
+//. ```javascript
+//. > parseArgs (spec) (Pair (conf) (['--email', 'xxx']))
+//. Left ('"xxx" is not a valid email address')
+//. ```
+//.
+//. Specifying an option multiple times is supported, and each handler defines
+//. its own semantics:
+//.
+//. ```javascript
+//. > parseArgs (spec) (Pair (conf) (['-c', '-c', '-c', '-c', '-c']))
+//. Right (Pair ({color: true, email: Nothing, words: []}) ([]))
+//.
+//. > parseArgs (spec) (Pair (conf) (['-e', 'a@x.org', '-e', 'b@x.org']))
+//. Right (Pair ({color: false, email: Just ('b@x.org'), words: []}) ([]))
+//.
+//. > parseArgs (spec) (Pair (conf) (['-w', 'foo', '-w', 'bar']))
+//. Right (Pair ({color: false, email: Nothing, words: ['foo', 'bar']}) ([]))
+//. ```
+//.
+//. As soon as an argument is encountered that is not a flag or an option
+//. value, that argument and any subsequent arguments are returned as the
+//. positional arguments (along with the final configuration value):
+//.
+//. ```javascript
+//. > parseArgs (spec) (Pair (conf) (['--color', 'foo', 'bar']))
+//. Right (Pair ({color: true, email: Nothing, words: []}) (['foo', 'bar']))
+//. ```
+//.
+//. `--` can be used to indicate that the remaining arguments are positional:
+//.
+//. ```javascript
+//. > parseArgs (spec) (Pair (conf) (['--color', '--', '--no-color']))
+//. Right (Pair ({color: true, email: Nothing, words: []}) (['--no-color']))
+//. ```
+//.
+//. An unrecognized flag/option name results in an error message:
+//.
+//. ```javascript
+//. > parseArgs (spec) (Pair (conf) (['-x', '-y', '-z']))
+//. Left ('-x is unrecognized')
+//.
+//. > parseArgs (spec) (Pair (conf) (['--xxx', '--yyy', '--zzz']))
+//. Left ('--xxx is unrecognized')
+//. ```
+//.
+//. Here is the type of `parseArgs` with type aliases expanded:
+//.
+//. ```haskell
+//. parseArgs :: StrMap (Either (a -> a) (String -> Either String (a -> a)))
+//.           -> Pair a (Array String)
+//.           -> Either String (Pair a (Array String))
+//. ```
+//.
+//. Although names can improve readability, the following example demonstrates
+//. that this module's functionality is exposed via a single, pure function:
+//.
+//. ```javascript
+//. > parseArgs
+//. .   ({'--foo': Left (conf => ({...conf, foo: true})),
+//. .     '--bar': Right (bar => Right (conf => ({...conf, bar: Just (bar)})))})
+//. .   (Pair ({foo: false, bar: Nothing})
+//. .         (['--foo', '--bar', 'baz', 'quux']))
+//. Right (Pair ({foo: true, bar: Just ('baz')}) (['quux']))
+//. ```
+const parseArgs =
+def ('parseArgs')
+    ({})
+    ([$.StrMap (Handler (a)),
+      $.Pair (a) ($.Array ($.String)),
+      $.Either ($.String) ($.Pair (a) ($.Array ($.String)))])
+    (spec => ([conf, args]) => recur (spec) (conf) (args));
+
+const recur = spec => conf => (
```

Comment:
The recursive implementation (although very elegant looking :drooling_face:) limits the length of arguments to at most the size of the stack. Would this be a problem for commands that take an option many (*many*) times? And would that limit be worth mentioning in the README?

Reply:
I consider this a reasonable limitation as it applies to flags and options but not to positional arguments. I can't imagine providing thousands of flags and options.

If someone *does* hit this limit, we can switch to an algorithm that is not recursive.

Reply:
Alright. It might be worth noting that it won't be the person who uses sanctuary-argv who hits that limit, but the user of their CLI.

For example, if a command takes `-f filename` multiple times, and me, as a user, do `find / -type f | intersperse -f | xargs their-cli {}+;` or something, then I'm the one who triggers the error, unbeknown to the author of `their-cli`.

I might be overthinking this though. I'd be happy to accept the limit myself, and just not create a CLI which provides any options that I expect will be repeated thousands of times.

Reply:
Your point about *who* might hit the limit is a very good one! I have updated the implementation. :)